ex2_system/source/diagnostic/diagnosic.c

```diff
@@ -20,38 +20,113 @@
 
 #include <FreeRTOS.h>
 #include <os_task.h>
+#include <os_semphr.h>
+#include "uhf.h"
+#include "eps.h"
+#include "system.h"
 
-static void *diagnostic_daemon(void *pvParameters);
+static void uhf_watchdog_daemon(void *pvParameters);
 SAT_returnState start_diagnostic_daemon(void);
 
+const unsigned int mutex_timeout = pdMS_TO_TICKS(100);
+static TickType_t prv_watchdog_delay = 1 * ONE_MINUTE; // 3 minutes
+
+static SemaphoreHandle_t uhf_watchdog_mtx = NULL;
 /**
- * Run subsystem diagnositc periodically at a low priority. Any corrective
- * action should be handled in the hardware interface layer.
+ * @brief Check that the UHF is responsive. If not, toggle power.
  *
- * @param pvParameters
- *    task parameters (not used)
+ * @param pvParameters Task parameters (not used)
  */
-static void *diagnostic_daemon(void *pvParameters) {
-    TickType_t delay = pdMS_TO_TICKS(1000);
+static void uhf_watchdog_daemon(void *pvParameters) {
+    TickType_t delay = prv_watchdog_delay;
     for (;;) {
-        // TODO Run subsystem diagnosics
+        // Get status word from UHF
+        char status[32];
+        const unsigned int retries = 3;
+        UHF_return err;
+        for (int i = 0; i < retries; i++) {
+            err = HAL_UHF_getSCW(status);
+            if (err == U_GOOD_CONFIG) {
+                break;
+            }
+        }
+
+        if (err != U_GOOD_CONFIG) {
+            ex2_log("UHF was not responsive - attempting to toggle power.");
+            // Toggle the UHF.
+            const unsigned int timeout = 5 * ONE_SECOND; // 30 seconds

```

Comment:
Merge master and resolve conflicts here. Remove or  update this comment to reflect the timeout that you want.

Reply:
This has been done.